File: app/src/main/java/com/colorgreen/swiping/MainActivity.java
package com.colorgreen.swiping;

import android.graphics.Color;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import com.colorgreen.swiper.OnSwipeTouchListener;
import com.colorgreen.swiper.SwipeAction;
import com.colorgreen.swiper.SwipeActionListener;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity {

    @BindView( R.id.main_layout )
    RelativeLayout mainLayout;
    @BindView( R.id.bar )
    LinearLayout bar;
    @BindView( R.id.bottombar )
    LinearLayout bottombar;
    @BindView( R.id.expand_button )
    Button expandButton;
    @BindView( R.id.collapse_button)
    Button collapseButton;

    @Override
    protected void onCreate( Bundle savedInstanceState ) {
        super.onCreate( savedInstanceState );
        setContentView( R.layout.activity_main );

        ButterKnife.bind( this );

        final int lightBlue = getColor( R.color.lightblue );
        final int darkBlue = getColor( R.color.darkblue );

        final OnSwipeTouchListener listener = new OnSwipeTouchListener();

        bar.getViewTreeObserver().addOnGlobalLayoutListener( new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                bar.getViewTreeObserver().removeOnGlobalLayoutListener( this );

                // for drawing layouts over buttons
                bar.setZ(10);
                bottombar.setZ(10);

                final int targetHeight = mainLayout.getHeight();

                final SwipeAction swipeAction = new SwipeAction();
                swipeAction.setDirection( SwipeAction.DragDirection.Down );
                swipeAction.setDragThreshold( 0.4f );
                swipeAction.setSteps( new float[]{ 0, targetHeight * 0.3f, targetHeight } );

                bottombar.setY( targetHeight );
                final SwipeAction bottomSwipeAction = new SwipeAction();
                bottomSwipeAction.setDirection( SwipeAction.DragDirection.Up );
                bottomSwipeAction.setDragThreshold( 0.2f );
                bottomSwipeAction.setSteps( new float[]{ targetHeight, targetHeight - targetHeight * 0.7f, 0 } );

                bottomSwipeAction.setSwipeActionListener( new SwipeActionListener() {
                    @Override
                    public void onDragStart( float val, float totalFriction ) {}

                    @Override
                    public void onDrag( float val, float totalFriction ) {
                        bottombar.setY( val );
                    }

                    @Override
                    public void onDragEnd( float val, float totalFriction ) {
                        swipeAction.setBlocked( bottomSwipeAction.isExtended() );
                    }
                } );

                swipeAction.setSwipeActionListener( new SwipeActionListener() {
                    @Override
                    public void onDragStart( float val, float totalFriction ) {}

                    @Override
                    public void onDrag( float val, float friction ) {
                        bar.setLayoutParams( new RelativeLayout.LayoutParams( bar.getWidth(), (int) val ) );
                        bar.setBackgroundColor( interpolateColor( lightBlue, darkBlue, val / targetHeight ) );
                    }

                    @Override
                    public void onDragEnd( float val, float totalFriction ) {
                        bottomSwipeAction.setBlocked( swipeAction.isExtended() );
                    }
                } );

                listener.addAction( swipeAction );
                listener.addAction( bottomSwipeAction );

                expandButton.setOnClickListener( v -> {
                    if( !swipeAction.isBlocked() )
                        swipeAction.expand();
                } );

                collapseButton.setOnClickListener( v -> {
                        if( !swipeAction.isBlocked() )
                            swipeAction.collapse();
                } );
            }
        } );

        listener.attachToView( mainLayout );
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    private float interpolate( float a, float b, float proportion ) {
        return ( a + ( ( b - a ) * proportion ) );
    }

    /**
     * Returns an interpoloated color, between <code>a</code> and <code>b</code>
     */
    private int interpolateColor( int a, int b, float proportion ) {
        float[] hsva = new float[3];
        float[] hsvb = new float[3];
        Color.colorToHSV( a, hsva );
        Color.colorToHSV( b, hsvb );
        for( int i = 0; i < 3; i++ ) {
            hsvb[i] = interpolate( hsva[i], hsvb[i], proportion );
        }
        return Color.HSVToColor( hsvb );
    }
}

File: app/src/main/java/com/colorgreen/swiping/MainActivity.kt
package com.colorgreen.swiping

import android.graphics.Color
import android.os.Bundle
import android.view.ViewTreeObserver
import android.widget.RelativeLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.colorgreen.swiper.OnSwipeTouchListener
import com.colorgreen.swiper.SwipeAction
import com.colorgreen.swiper.SwipeActionListener
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val listener = OnSwipeTouchListener()

        bar.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                bar.viewTreeObserver.removeOnGlobalLayoutListener(this)

                val targetHeight = main_layout.height
                // for drawing layouts over buttons
                bar.z = 10f
                bottombar.y = main_layout.height.toFloat()
                bottombar.z = 10f

                val swipeAction = SwipeAction(
                    SwipeAction.DragDirection.Down,
                    floatArrayOf(0f, targetHeight * 0.3f, targetHeight.toFloat()),
                    0.4f
                )
                val bottomSwipeAction = SwipeAction(
                    SwipeAction.DragDirection.Up,
                    floatArrayOf(targetHeight.toFloat(), targetHeight - targetHeight * 0.7f, 0f),
                    0.2f
                )

                bottomSwipeAction.setSwipeActionListener(object : SwipeActionListener {
                    override fun onDragStart(value: Float, totalFriction: Float) {}

                    override fun onDrag(value: Float, totalFriction: Float) {
                        bottombar.y = value
                    }

                    override fun onDragEnd(value: Float, totalFriction: Float) {
                        swipeAction.isBlocked = bottomSwipeAction.isExtended()
                    }
                })

                swipeAction.setSwipeActionListener(object : SwipeActionListener {
                    override fun onDragStart(value: Float, totalFriction: Float) {}

                    override fun onDrag(value: Float, totalFriction: Float) {
                        bar!!.layoutParams = RelativeLayout.LayoutParams(bar!!.width, value.toInt())
                        bar!!.setBackgroundColor(
                            interpolateColor(
                                ContextCompat.getColor(this@MainActivity, R.color.lightblue),
                                ContextCompat.getColor(this@MainActivity, R.color.darkblue),
                                value / targetHeight
                            )
                        )
                    }

                    override fun onDragEnd(value: Float, totalFriction: Float) {
                        bottomSwipeAction.isBlocked = swipeAction.isExtended()
                    }
                })

                listener.addAction(swipeAction)
                listener.addAction(bottomSwipeAction)

                expand_button.setOnClickListener {
                    if (!swipeAction.isBlocked)
                        swipeAction.expand()
                }

                collapse_button.setOnClickListener {
                    if (!swipeAction.isBlocked)
                        swipeAction.collapse()
                }
            }
        })

        listener.attachToView(main_layout)
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    private fun interpolate(a: Float, b: Float, proportion: Float): Float {
        return a + (b - a) * proportion
    }

    /**
     * Returns an interpoloated color, between `a` and `b`
     */
    private fun interpolateColor(a: Int, b: Int, proportion: Float): Int {
        val hsva = FloatArray(3)
        val hsvb = FloatArray(3)
        Color.colorToHSV(a, hsva)
        Color.colorToHSV(b, hsvb)
        for (i in 0..2) {
            hsvb[i] = interpolate(hsva[i], hsvb[i], proportion)
        }
        return Color.HSVToColor(hsvb)
    }
}
